File: functions/audiofile.py
import numpy as np
import soundfile as sf
from scipy.signal import butter, sosfiltfilt, iirnotch, tf2sos
from .blanket import Blanket


class AudioFile:
    def __init__(self, path: str):
        self.data, self.sample_rate = sf.read(path, always_2d=True)
        self.path = path
        self.blanket = Blanket(self)

    def save(self, path: str = None) -> "AudioFile":
        sf.write(path or self.path, self.data, self.sample_rate)
        return self

    # --- Filters ---

    def high_pass(self, cutoff_hz: float, order: int = 4) -> "AudioFile":
        sos = butter(order, cutoff_hz, btype="high", fs=self.sample_rate, output="sos")
        self.data = sosfiltfilt(sos, self.data, axis=0)
        return self

    def low_pass(self, cutoff_hz: float, order: int = 4) -> "AudioFile":
        sos = butter(order, cutoff_hz, btype="low", fs=self.sample_rate, output="sos")
        self.data = sosfiltfilt(sos, self.data, axis=0)
        return self

    def band_pass(self, low_hz: float, high_hz: float, order: int = 4) -> "AudioFile":
        sos = butter(order, [low_hz, high_hz], btype="band", fs=self.sample_rate, output="sos")
        self.data = sosfiltfilt(sos, self.data, axis=0)
        return self

    def notch(self, center_hz: float, q: float = 30.0) -> "AudioFile":
        """Narrow cut at a specific frequency. Higher Q = narrower."""
        b, a = iirnotch(center_hz, q, self.sample_rate)
        sos = tf2sos(b, a)
        self.data = sosfiltfilt(sos, self.data, axis=0)
        return self

    # --- Parametric / Shelving EQ (Audio EQ Cookbook biquads) ---

    def peak_eq(self, center_hz: float, gain_db: float, q: float) -> "AudioFile":
        """Boost or cut a frequency band. Negative gain_db cuts."""
        A = 10 ** (gain_db / 40)
        w0 = 2 * np.pi * center_hz / self.sample_rate
        alpha = np.sin(w0) / (2 * q)
        b = np.array([1 + alpha * A, -2 * np.cos(w0), 1 - alpha * A])
        a = np.array([1 + alpha / A, -2 * np.cos(w0), 1 - alpha / A])
        self.data = sosfiltfilt(tf2sos(b, a), self.data, axis=0)
        return self

    def low_shelf(self, cutoff_hz: float, gain_db: float, slope: float = 1.0) -> "AudioFile":
        """Boost or cut all frequencies below cutoff_hz."""
        A = 10 ** (gain_db / 40)
        w0 = 2 * np.pi * cutoff_hz / self.sample_rate
        alpha = np.sin(w0) / 2 * np.sqrt((A + 1 / A) * (1 / slope - 1) + 2)
        cos_w0 = np.cos(w0)
        sqrt_A = np.sqrt(A)
        b = np.array([
            A * ((A + 1) - (A - 1) * cos_w0 + 2 * sqrt_A * alpha),
            2 * A * ((A - 1) - (A + 1) * cos_w0),
            A * ((A + 1) - (A - 1) * cos_w0 - 2 * sqrt_A * alpha),
        ])
        a = np.array([
            (A + 1) + (A - 1) * cos_w0 + 2 * sqrt_A * alpha,
            -2 * ((A - 1) + (A + 1) * cos_w0),
            (A + 1) + (A - 1) * cos_w0 - 2 * sqrt_A * alpha,
        ])
        self.data = sosfiltfilt(tf2sos(b, a), self.data, axis=0)
        return self

    def high_shelf(self, cutoff_hz: float, gain_db: float, slope: float = 1.0) -> "AudioFile":
        """Boost or cut all frequencies above cutoff_hz."""
        A = 10 ** (gain_db / 40)
        w0 = 2 * np.pi * cutoff_hz / self.sample_rate
        alpha = np.sin(w0) / 2 * np.sqrt((A + 1 / A) * (1 / slope - 1) + 2)
        cos_w0 = np.cos(w0)
        sqrt_A = np.sqrt(A)
        b = np.array([
            A * ((A + 1) + (A - 1) * cos_w0 + 2 * sqrt_A * alpha),
            -2 * A * ((A - 1) + (A + 1) * cos_w0),
            A * ((A + 1) + (A - 1) * cos_w0 - 2 * sqrt_A * alpha),
        ])
        a = np.array([
            (A + 1) - (A - 1) * cos_w0 + 2 * sqrt_A * alpha,
            2 * ((A - 1) - (A + 1) * cos_w0),
            (A + 1) - (A - 1) * cos_w0 - 2 * sqrt_A * alpha,
        ])
        self.data = sosfiltfilt(tf2sos(b, a), self.data, axis=0)
        return self

    # --- Gain ---

    def gain(self, db: float) -> "AudioFile":
        self.data = self.data * (10 ** (db / 20))
        return self

File: mobile/src/dsp/AudioFile.ts
// Port of functions/audiofile.py. Holds decoded PCM + the deferred effect
// chain, and exposes the raw DSP primitives. Methods append a spec and return
// `this` for chaining, mirroring the Python's mutate-and-return style.

import { Platform } from 'react-native';
import { OfflineAudioContext, decodeAudioData } from 'react-native-audio-api';
import { Blanket } from './Blanket';
import { Channels, EffectSpec, BiquadKind } from './types';
import { renderChain, RenderResult } from './render';

// All decode/render runs at this rate, so PCM held in the session can be
// rebuilt into an AudioFile without separately tracking its sample rate.
export const TARGET_SAMPLE_RATE = 48000;

// Butterworth section Q values per even order (cascaded biquads).
const BUTTER_Q: Record<number, number[]> = {
  2: [0.7071],
  4: [0.5412, 1.3066],
  6: [0.5176, 0.7071, 1.9319],
  8: [0.5098, 0.6013, 0.9000, 2.5629],
};

export class AudioFile {
  data: Channels;
  sampleRate: number;
  chain: EffectSpec[] = [];
  blanket: Blanket;

  constructor(data: Channels, sampleRate: number) {
    this.data = data;
    this.sampleRate = sampleRate;
    this.blanket = new Blanket(this);
  }

  /** Decode an audio/video file (mp4/m4a/mp3/wav) to PCM. */
  static async decode(uri: string, targetSampleRate = TARGET_SAMPLE_RATE): Promise<AudioFile> {
    let buffer;
    if (Platform.OS === 'web') {
      // The browser decoder only takes an ArrayBuffer, so fetch the blob URL's
      // bytes first. (Passing the URL string straight in throws on web.)
      const ctx = new OfflineAudioContext({
        numberOfChannels: 2,
        length: 1,
        sampleRate: targetSampleRate,
      });
      const bytes = await fetch(uri).then((r) => r.arrayBuffer());
      buffer = await ctx.decodeAudioData(bytes);
    } else {
      // Native: decode straight from the file path. This goes to the native
      // file-path decoder, skipping the throwaway OfflineAudioContext.
      buffer = await decodeAudioData(uri, targetSampleRate);
    }

    const channels: Channels = [];
    for (let c = 0; c < buffer.numberOfChannels; c++) {
      // slice() is a memcpy; Float32Array.from() copies element-by-element.
      channels.push(buffer.getChannelData(c).slice());
    }
    return new AudioFile(channels, buffer.sampleRate);
  }

  /** Append a spec to the chain. */
  add(spec: EffectSpec): this {
    this.chain.push(spec);
    return this;
  }

  /** Execute the whole chain offline and return processed channels. */
  render(): Promise<RenderResult> {
    return renderChain(this.data, this.sampleRate, this.chain);
  }

  // --- Filters ---

  private butterworth(kind: BiquadKind, frequency: number, order: number): this {
    const qs = BUTTER_Q[order] ?? BUTTER_Q[4];
    for (const q of qs) this.add({ kind, frequency, q });
    return this;
  }

  high_pass(cutoffHz: number, order = 4): this {
    return this.butterworth('highpass', cutoffHz, order);
  }

  low_pass(cutoffHz: number, order = 4): this {
    return this.butterworth('lowpass', cutoffHz, order);
  }

  band_pass(lowHz: number, highHz: number): this {
    const frequency = Math.sqrt(lowHz * highHz);
    const q = frequency / (highHz - lowHz);
    return this.add({ kind: 'bandpass', frequency, q });
  }

  notch(centerHz: number, q = 30.0): this {
    return this.add({ kind: 'notch', frequency: centerHz, q });
  }

  // --- Parametric / Shelving EQ ---

  peak_eq(centerHz: number, gainDb: number, q: number): this {
    return this.add({ kind: 'peaking', frequency: centerHz, q, gainDb });
  }

  low_shelf(cutoffHz: number, gainDb: number): this {
    return this.add({ kind: 'lowshelf', frequency: cutoffHz, gainDb });
  }

  high_shelf(cutoffHz: number, gainDb: number): this {
    return this.add({ kind: 'highshelf', frequency: cutoffHz, gainDb });
  }

  // --- Gain ---

  gain(db: number): this {
    return this.add({ kind: 'gain', db });
  }
}
